app/actions/websocket/roles.ts

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {fetchRolesIfNeeded, RolesRequest} from '@actions/remote/role';
+import {fetchMe} from '@actions/remote/user';
+import DatabaseManager from '@database/manager';
+import {queryRoleById} from '@queries/servers/role';
+import {queryCurrentUserId} from '@queries/servers/system';
+import {prepareMyTeams} from '@queries/servers/team';
+import {prepareUsers} from '@queries/servers/user';
+import {WebSocketMessage} from '@typings/api/websocket';
+
+import type {Model} from '@nozbe/watermelondb';
+import type {ServerDatabase} from '@typings/database/database';
+import type RoleModel from '@typings/database/models/servers/role';
+
+export async function handleRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    // only update Role records that exist in the Role Table
+    const role = JSON.parse(msg.data.role);
+    const dbRole = await queryRoleById(database.database, role.id);
+    if (!dbRole) {
+        return;
+    }
+
+    await database.database.write(async () => {
+        await dbRole.update((roleRecord: RoleModel) => {
+            roleRecord.permissions = role.permissions;
+        });
+    });
+}
+
+export async function handleUserRoleUpdatedEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
+
+    const currentUserId = await queryCurrentUserId(database.database);
+    if (currentUserId !== msg.data.user_id) {
+        return;
+    }
+
+    const modelPromises: Array<Promise<Model[]>> = [];
+
+    // update Role Table if needed
+    const newRoles = await fetchRolesIfNeeded(serverUrl, Array.from(msg.data.roles), true);
+    const preparedRoleModels = getPreparedRoleModels(database, newRoles);
+    modelPromises.push(preparedRoleModels);
+
+    // update User Table record
+    const meData = await fetchMe(serverUrl, true);
```

Comment:
The WS message does not include a full server model user. It only includes `user_id` and `roles`. 

In this case, we have a couple of options. 
1. retrieve user from server with `fetchMe` (current implementation)
  1. pass server User model to `prepareUsers`
2. query the app db UserModel, modify and write it back
3. write a patching function for patching the roles field of a UserModel without concern for other fields not changing.

I leaned toward #1 so so that the actual user from the server model is exact with modifying the User Table. This shouldn't be a concern that other fields (besides Roles) would change, but this insures the user model is latest from the server.

Reply:
I think we should go with number 3. A call to the database is less costly than a call to the server. You can use the `prepareUpdate` function of the model. You can find an example of this on `app/actions/local/channel.ts`.

Reply:
Done.  Thanks for pointing out the channel example